File: internal/tui/tui.go
package tui

import (
	"bufio"
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"
)

const (
	CReset  = "\033[0m"
	CInfo   = "\033[36m"   // Cyan
	COk     = "\033[32m"   // Green
	CWarn   = "\033[33m"   // Yellow
	CErr    = "\033[31m"   // Red
	CAccent = "\033[35m"   // Magenta
)

func PrintHeader(title string) {
	fmt.Printf("\n%s=== %s ===%s\n", CAccent, title, CReset)
}

func PrintInfo(msg string) {
	fmt.Printf("  %s[INFO]%s %s\n", CInfo, CReset, msg)
}

func PrintOk(msg string) {
	fmt.Printf("  %s[OK]%s   %s\n", COk, CReset, msg)
}

func PrintWarn(msg string) {
	fmt.Printf("  %s[WARN]%s %s\n", CWarn, CReset, msg)
}

func PrintErr(msg string) {
	fmt.Printf("  %s[ERR]%s  %s\n", CErr, CReset, msg)
}

func PrintKV(key, val string) {
	fmt.Printf("  %s%-20s%s %s\n", CInfo, key, CReset, val)
}

// PromptMenu displays a numbered menu and prompts for selection (0-indexed).
// If defaultIdx >= 0 and user inputs empty string, defaultIdx is returned.
func PromptMenu(title string, options []string, defaultIdx int) int {
	PrintHeader(title)
	for i, opt := range options {
		suffix := ""
		if defaultIdx >= 0 && i == defaultIdx {
			suffix = " (Mac dinh)"
		}
		fmt.Printf("  %d) %s%s\n", i+1, opt, suffix)
	}

	hint := ""
	if defaultIdx >= 0 {
		hint = fmt.Sprintf(" [Mac dinh: %d]", defaultIdx+1)
	}

	scanner := bufio.NewScanner(os.Stdin)
	for {
		fmt.Printf("  Nhap lua chon (1-%d)%s: ", len(options), hint)
		if !scanner.Scan() {
			fmt.Println("\n  [INFO] Huy bo boi nguoi dung.")
			os.Exit(130)
		}

		input := strings.TrimSpace(scanner.Text())
		if input == "" {
			if defaultIdx >= 0 {
				return defaultIdx
			}
			continue
		}

		val, err := strconv.Atoi(input)
		if err != nil {
			PrintWarn("Vui long nhap mot so.")
			continue
		}

		idx := val - 1
		if idx >= 0 && idx < len(options) {
			return idx
		}
		PrintWarn(fmt.Sprintf("Lua chon khong hop le. Nhap tu 1 den %d.", len(options)))
	}
}

type ChecklistItem struct {
	Key     string
	Label   string
	Checked bool
}

// PromptChecklist displays toggleable items.
func PromptChecklist(title string, items []ChecklistItem) []string {
	selected := make([]bool, len(items))
	for i, item := range items {
		selected[i] = item.Checked
	}

	scanner := bufio.NewScanner(os.Stdin)
	for {
		PrintHeader(title)
		for i, item := range items {
			status := "[ ]"
			if selected[i] {
				status = "[X]"
			}
			fmt.Printf("  %s %d) %s\n", status, i+1, item.Label)
		}
		fmt.Println("\n  Nhap so de chon/bo chon (vi du: '1 2'), hoac an Enter de xac nhan.")
		fmt.Print("  Lua chon: ")

		if !scanner.Scan() {
			fmt.Println("\n  [INFO] Huy bo boi nguoi dung.")
			os.Exit(130)
		}

		input := strings.TrimSpace(scanner.Text())
		if input == "" {
			break
		}

		fields := strings.Fields(input)
		for _, f := range fields {
			if idx, err := strconv.Atoi(f); err == nil {
				idx = idx - 1
				if idx >= 0 && idx < len(items) {
					selected[idx] = !selected[idx]
				}
			}
		}
	}

	var result []string
	for i, sel := range selected {
		if sel {
			result = append(result, items[i].Key)
		}
	}
	return result
}

func getSystemEditor() string {
	editor := os.Getenv("EDITOR")
	if editor != "" {
		return editor
	}
	for _, candidate := range []string{"nano", "vim", "vi"} {
		if _, err := exec.LookPath(candidate); err == nil {
			return candidate
		}
	}
	return "nano"
}

// EditTextInEditor launches system editor on a temp file with initial content and returns updated content.
func EditTextInEditor(content, suffix string) (string, error) {
	tempFile, err := os.CreateTemp("", "*"+suffix)
	if err != nil {
		return content, err
	}
	tempPath := tempFile.Name()
	defer os.Remove(tempPath)

	if _, err := tempFile.WriteString(content); err != nil {
		tempFile.Close()
		return content, err
	}
	tempFile.Close()

	editor := getSystemEditor()
	cmd := exec.Command(editor, tempPath)
	cmd.Stdin = os.Stdin
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	if err := cmd.Run(); err != nil {
		PrintErr(fmt.Sprintf("Editor %s gap loi. Su dung noi dung mac dinh.", editor))
		return content, err
	}

	data, err := os.ReadFile(tempPath)
	if err != nil {
		return content, err
	}
	return string(data), nil
}

// OpenFileInEditor opens a file with the system editor.
func OpenFileInEditor(filePath string) error {
	editor := getSystemEditor()
	cmd := exec.Command(editor, filePath)
	cmd.Stdin = os.Stdin
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	return cmd.Run()
}

File: xsync_core/tui.py
import os
import sys
import tempfile
import subprocess


C_RESET = '\033[0m'
C_INFO = '\033[36m'   # Cyan
C_OK = '\033[32m'     # Green
C_WARN = '\033[33m'   # Yellow
C_ERR = '\033[31m'    # Red
C_ACCENT = '\033[35m' # Magenta

def print_header(title):
    print(f"\n{C_ACCENT}=== {title} ==={C_RESET}")

def print_info(msg):
    print(f"  {C_INFO}[INFO]{C_RESET} {msg}")

def print_ok(msg):
    print(f"  {C_OK}[OK]{C_RESET}   {msg}")

def print_warn(msg):
    print(f"  {C_WARN}[WARN]{C_RESET} {msg}")

def print_err(msg):
    print(f"  {C_ERR}[ERR]{C_RESET}  {msg}")

def print_kv(key, val):
    print(f"  {C_INFO}{key:<20}{C_RESET} {val}")

def prompt_menu(title, options, default_idx=None):
    """
    Displays a numbered menu and prompts for selection.
    If default_idx is provided and the user presses Enter, returns default_idx.
    """
    print_header(title)
    for idx, opt in enumerate(options, 1):
        suffix = " (Mac dinh)" if default_idx is not None and idx - 1 == default_idx else ""
        print(f"  {idx}) {opt}{suffix}")
    
    hint = f" [Mac dinh: {default_idx + 1}]" if default_idx is not None else ""
    while True:
        try:
            choice = input(f"  Nhap lua chon (1-{len(options)}){hint}: ").strip()
            if not choice:
                if default_idx is not None:
                    return default_idx
                continue
            choice_idx = int(choice) - 1
            if 0 <= choice_idx < len(options):
                return choice_idx
            print_warn(f"Lua chon khong hop le. Nhap tu 1 den {len(options)}.")
        except ValueError:
            print_warn("Vui long nhap mot so.")
        except KeyboardInterrupt:
            print("\n  [INFO] Huy bo boi nguoi dung.")
            sys.exit(130)

def prompt_checklist(title, items):
    """
    Items is a list of tuples: (key, label, default_on_boolean)
    Returns list of selected keys.
    """
    selected = [item[2] for item in items]
    while True:
        print_header(title)
        for idx, item in enumerate(items, 1):
            status = "[X]" if selected[idx-1] else "[ ]"
            print(f"  {status} {idx}) {item[1]}")
        print("\n  Nhap so de chon/bo chon (vi du: '1 2'), hoac an Enter de xac nhan.")
        try:
            choice = input("  Lua chon: ").strip()
            if not choice:
                break
            parts = choice.split()
            for p in parts:
                if p.isdigit():
                    idx = int(p) - 1
                    if 0 <= idx < len(items):
                        selected[idx] = not selected[idx]
        except KeyboardInterrupt:
            print("\n  [INFO] Huy bo boi nguoi dung.")
            sys.exit(130)
    return [items[i][0] for i, sel in enumerate(selected) if sel]



def edit_text_in_editor(content, filename_suffix="_whitelist"):
    """
    Launches editor on a temp file with content and returns the updated content.
    """
    # Create temp file
    fd, temp_path = tempfile.mkstemp(suffix=filename_suffix, text=True)
    try:
        with os.fdopen(fd, 'w', encoding='utf-8') as f:
            f.write(content)
            
        # Find editor
        editor = os.environ.get("EDITOR", "")
        if not editor:
            # Check availability
            for candidate in ["nano", "vim", "vi"]:
                if subprocess.run(["which", candidate], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL).returncode == 0:
                    editor = candidate
                    break
        if not editor:
            editor = "nano" # fallback
            
        # Run editor
        res = subprocess.run([editor, temp_path])
        if res.returncode != 0:
            print_err(f"Editor {editor} error. Dat lai mac dinh.")
            return content
            
        # Read back
        with open(temp_path, 'r', encoding='utf-8') as f:
            updated = f.read()
        return updated
    finally:
        if os.path.exists(temp_path):
            os.remove(temp_path)

def open_file_in_editor(file_path):
    """
    Opens a file directly using the system editor.
    """
    editor = os.environ.get("EDITOR", "")
    if not editor:
        for candidate in ["nano", "vim", "vi"]:
            if subprocess.run(["which", candidate], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL).returncode == 0:
                editor = candidate
                break
    if not editor:
        editor = "nano"
    subprocess.run([editor, file_path])
